Hw2

IAGrig opened the pull request.

IAGrig committed:
build: add database module and prettify initial configs

IAGrig committed:
feat: implement bot and all commands, add allowed domains

The bot starts with Spring, gets token from enviromental variables.
Command processing is based on Chain of Responsibility pattern.
The LinksHandler is responsible for processing and validating
incoming links for tracking.
The allowed domains are specified in the configuration files as
a single line separated by commas without spaces.

IAGrig committed:
feat: add database support

The Stub Database is a temporal in-memory database, which will be
replaced by a real database in the future.

IAGrig committed:
test: add tests for commands handling and LinksHandler

IAGrig committed:
feat: add a menu with commands to the bot and do a little refactoring

IAGrig committed:
fix: fix a bug in /list (links id were not displayed)

The user must to specify the link id to stop tracking the link using
/untrack, but these ids were not displayed in /list output.

IAGrig committed:
feat: implement GitHub and Stackoverflow HTTP clients

In addition, add DTO classes in java and baseUrls in application.yaml
for GitHub and Stackoverflow.

IAGrig committed:
test: add tests for GitHubHttpClient and StackoverflowHttpClient

IAGrig committed:
build: fix wiremock dependencies

IAGrig committed:
feat: add a scheduled method, register it

IAGrig committed:
fix: correct all defects

The application configuration has been moved from the dev profile
to the general one. Unused fields in classes have been removed.
Fixed the processing of the /start command and the processing of
allowed domains, added functionality for tracking registered users.

IAGrig committed:
test: fix the tests - replace junit4 with junit5

damir2407 commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -0,0 +1,26 @@
+
+@Configuration
+public class ClientConfiguration {
+    @Value("${scrapper.stackoverflow.baseUrl}")
```
Давай прям сюда и напишем урл по умолчанию? 
Что-то типа:     @Value("${scrapper.stackoverflow.baseUrl:https://......}")


> При создании клиента обязательно должна быть возможность указать базовый URL. При этом если он не указывается, то должен использоваться URL по умолчанию.


А то у тебя это не сделано

damir2407 commented on scrapper/src/main/java/edu/java/scrapper/github/GithubHttpClient.java:
```diff
@@ -0,0 +1,55 @@
+    private String baseUrlDefault;
+
+    public GithubHttpClient(String baseUrl) {
+        client = getWebClientWithBaseUrl(baseUrl);
```
Давай бин WebClient'а внедрим там же где ты и создаешь бины клиентов для гитхаба и стаковерфлоу?

damir2407 commented on scrapper/src/main/java/edu/java/scrapper/github/GithubHttpClient.java:
```diff
@@ -0,0 +1,55 @@
+import org.springframework.stereotype.Component;
+import org.springframework.web.reactive.function.client.WebClient;
+
+@Component
```
Ты внедрил клиентов как бины через конф класс. Тебе не нужен @Component

damir2407 commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -0,0 +1,26 @@
+    private String baseGithubUrlDefault;
+
+    @Bean
+    public GithubHttpClient githubHttpClient() {
```
Внедряй используя родителя, то-есть HttpClient, вместо GithubHttpClient и также для стаковерфлоу

damir2407 commented on scrapper/src/main/java/edu/java/scrapper/stackoverflow/StackoverflowHttpClient.java:
```diff
@@ -0,0 +1,60 @@
+import org.springframework.web.reactive.function.client.WebClient;
+import org.springframework.web.util.UriBuilder;
+
+@Component
```
Ты внедрил клиентов как бины через конф класс. Тебе не нужен @component

damir2407 commented on scrapper/src/main/java/edu/java/scrapper/LinkUpdateScheduler.java:
```diff
@@ -0,0 +1,19 @@
+
+@Log4j2
+@Component
+@EnableScheduling
```
@EnableScheduling пишется над main классом

IAGrig committed:
fix: correct all defects

Add default values for URL paths. Move the creation of WebClients
to ClientConfiguration. Move @EnableScheduling to the main class.
Fix the lack of 404 error handling in some Http Clients methods.

IAGrig committed:
test: fix the tests - replace junit4 with junit5

IAGrig committed:
Merge branch 'hw1' into hw3

damir2407 commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -0,0 +1,54 @@
+    private String baseGithubUrlDefault;
+
+    @Bean
+    @Qualifier("githubHttpClient")
```
@Qualifier ставится при инжекте бина. При создании бина лишь указывается его имя. При создании через @Bean будет @Bean(name = "githubHttpClient")

damir2407 replied:
Если хочешь задать имя бина, то оно задается через @Bean(name = "name"). @Qualifier при создании бина не гарантирует его уникальности, это значит, что у тебя может быть несколько одинаковых квалифаеров, в то время как бин нейм должен быть уникален

damir2407 commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Bean
+    @Qualifier("githubHttpClient")
+    public HttpClient githubHttpClient(WebClient githubWebClient) {
```
вот у тебя допустим githubWebClient, но ты явно не говоришь какой бин внедрить, хотя у тебя несколько бинов с одинаковым типом (гитхаб и стаковерфлоу); Вот сюда тебе нужно сделать:

`    public HttpClient githubHttpClient(@Qualifier("githubWebClient") WebClient githubWebClient) {
`

damir2407 commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Bean
+    @Qualifier("stackoverflowHttpClient")
```
та же история, давай уберем qulifier'ы, у тебя и так имена бинов в названии методов заданы и они адекватны

damir2407 replied:
дефолтная стратегия названия бина - его имя, в твоем случае название метода

damir2407 commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Bean
+    @Qualifier("stackoverflowHttpClient")
+    public HttpClient stackoverflowHttpClient(WebClient stackoverflowWebClient) {
```
та же история

damir2407 commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -0,0 +1,54 @@
+        );
+    }
+
+    @Bean
```
ну и сюда соответственно вписать имя бина. то-есть @Bean("githubWebClient")

damir2407 commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -0,0 +1,54 @@
+            .build();
+    }
+
+    @Bean
```
та же история

IAGrig committed:
feat(bot): add a method for sending messages to the user

IAGrig committed:
feat(database): add a method for deleting users and links

IAGrig committed:
feat(bot): add API controller, DTOs, service, exception handler

IAGrig committed:
build(scrapper): add database dependency to scrapper

IAGrig committed:
feat(scrapper): add API controller, DTOs, services, exception handler

IAGrig committed:
refactor: tiny refactoring

IAGrig committed:
build: add swagger-ui

IAGrig committed:
refactor: separate DTOs into the models module

IAGrig committed:
refactor: move http clients to a separate package

IAGrig committed:
feat(bot): add http client for scrapper

IAGrig committed:
feat(scrapper): add http client for bot

IAGrig committed:
fix(models): make DTOs to use URIs for links

IAGrig committed:
fix: make track/untrack endpoints return LinkResponse

IAGrig committed:
feat: add advanced exception handling

IAGrig committed:
fix(scrapper): delete useless @Qualifier, set a name for the beans

damir2407 commented on scrapper/src/main/java/edu/java/configuration/ClientConfiguration.java:
```diff
@@ -20,14 +19,12 @@ public class ClientConfiguration {
 
-    @Bean
-    @Qualifier("githubHttpClient")
+    @Bean(name = "githubHttpClient")
```
Ставить name тоже не имеет смысла в твоем случае) У тебя метод называется githubHttpClient -> бин создастся с этим именем. Провались для интереса в аннотацию @Bean и почитай про это

IAGrig replied:
Мне вообще убрать название или сделать @Bean("githubHttpClient")? Вроде бы без названия не будет работать. Я пытался писать спринговые тесты с @Autowire и там была с этим проблема. Несколько бинов HttpClient, нужно уточнить необходимый. 

Кажется дошло. Можно было просто там добавить @Qualifier с именем метода, а тут ничего не менять?

damir2407 replied:
> Кажется дошло. Можно было просто там добавить @qualifier с именем метода, а тут ничего не менять?

Да)

IAGrig committed:
fix(scrapper): remove useless names for beans

IAGrig committed:
fix: remove useless @JsonProperty, move endpoints to controllers level

Move /updates to the controller level in UpdatesController
Move /tg-chat to the controller level in TgChatConroller
Move /links to the controller level in LinksController

IAGrig committed:
fix: register http clients as beans, add configuration for them

IAGrig committed:
test: add jdbc dependencies

IAGrig committed:
feat: add liquibase migrations

IAGrig committed:
test: add the testcontainer integration test

IAGrig committed:
fix: fix a bug in ScrapperHttpClient, fix URI validation

IAGrig committed:
Merge branch 'hw4' into hw5, move database tests to separate package

IAGrig committed:
refactor: move command handlers to stub package, add User, updade Link

IAGrig committed:
feat: add JDBC repositories for links and users

IAGrig committed:
test: add test for JDBC link and user repositories

IAGrig committed:
feat: add JDBC services, rewrite controllers, add command handlers

IAGrig committed:
feat: implement scheduled changes checker, connect scrapper and bot

IAGrig committed:
fix: fix URI validation

IAGrig committed:
feat: add a detailed update notifications for Stackoverflow

IAGrig committed:
feat: make my plain java entities real persistence entities

IAGrig committed:
feat: add JPA repositories and services

IAGrig committed:
test: add JPA services and repositories tests

IAGrig committed:
feat: implement a configurable database backend selection

IAGrig committed:
fix: add a test configuration for JPA services tests, fix repo tests

IAGrig committed:
refactor: separate link checkers to beans

IAGrig committed:
refactor: make simple beans components

IAGrig committed:
feat: add retry managers to all HTTP clients

IAGrig committed:
feat: add retry codes and linear retry

IAGrig committed:
feat(bot): add bucket4j rate limit for public endpoints

IAGrig committed:
feat(scrapper): add bucket4j rate limit for public endpoints

IAGrig committed:
fix: add missed and remove useless annotations, add Optionals to JPA

IAGrig committed:
feat: add kafka and zookeeper to the compose.yml

IAGrig committed:
feat(scrapper): add kafka properties, producers and updates handler

IAGrig committed:
feat(bot): add kafka properties and listener service

IAGrig committed:
fix(scrapper): add missing annotations, fix conditional bean creation

IAGrig committed:
fix(bot): fix configs, add updates handling and missing annotation

IAGrig committed:
feat(bot): implement Dead Letter Queue

IAGrig committed:
fix(scrapper): add null check in update method

IAGrig committed:
test(scrapper): add tests to the scrapper kafka producer

IAGrig committed:
feat(bot): add metrics and micrometer dependencies

IAGrig committed:
feat(scrapper): add metrics and micrometer dependencies

IAGrig committed:
feat: add prometheus config, grafana and prometheus images to the compose

IAGrig committed:
feat(bot): add a custom metric - count of processed updates

IAGrig committed:
feat(scrapper): add a custom metric - count of processed updates

IAGrig committed:
fix(bot): change bot handling of null tgIds, remove a partial sending

IAGrig committed:
fix: fix broken tests, add correct tests to bot kafka consumer

IAGrig committed:
fix(scrapper): fix the processing of adding already tracked links

IAGrig committed:
test(scrapper): fix problem with same user IDs in parallel tests

IAGrig committed:
feat(bot): GitHub actions for publishing to the GitHub Container Registry

IAGrig committed:
feat(scrapper): GitHub actions for publishing to the GitHub Registry

IAGrig committed:
feat(scrapper): add JOOQ to the configuration

IAGrig committed:
feat(scrapper): add JOOQ codegen and generated classes

IAGrig committed:
feat(scrapper): implement JOOQ repositories and services

IAGrig committed:
test(scrapper): add tests for JOOQ services and repositories

IAGrig committed:
Merge branch 'hw5-bonus-jooq' into hw10

IAGrig committed:
Merge branch 'hw10' into hw9

IAGrig committed:
Merge branch 'hw9' into hw8

IAGrig committed:
Merge branch 'hw8' into hw7

IAGrig committed:
Merge branch 'hw7' into hw6

IAGrig committed:
Merge branch 'hw6' into hw5-bonus

IAGrig committed:
fix(scrapper): remove unnecessary @Qualifier annotations

IAGrig committed:
Merge branch 'hw5-bonus' into hw5

IAGrig committed:
Merge branch 'hw5' into hw3

IAGrig committed:
Merge branch 'hw3' into hw2
